opm/autodiff/FlowMain.hpp

```diff
@@ -252,9 +253,67 @@ namespace Opm
 #endif
         }
 
+        /// checks cartesian adjacency of global indices g1 and g2
+        bool cartesianAdjacent(const Grid& grid, size_t g1, size_t g2) {
+            // we need cartDims from UgGridHelpers
+            using namespace UgGridHelpers;
 
+            auto diff = std::abs(g1 - g2);
 
+            const int * dimens = cartDims(grid);
+            if (diff == 1)
+               return true;
+            if (diff == dimens[0])
+               return true;
+            if (diff == dimens[0] * dimens[1])
+               return true;
 
+            return false;
+        }
+
+        /*!
+         * \brief Get the NNC structure of the given grid.
+         *
+         * Get cell adjacencies beyond Cartesian neighborhoods.  The
+         * transmissivity is given as -1d (unless trans vector is provided)
+         * hence NNC "structure".
+         *
+         * The trans vector is indexed by face number as it is in grid.
+         */
+        NNC nncStructure(const Grid& grid, const std::vector<double>& trans = std::vector<double>()) {
+            // we use numFaces, numCells, cell2Faces, globalCell from UgGridHelpers
+            using namespace UgGridHelpers;
+            NNC nnc;
+
+            typedef std::vector<int> nb; // neighbors
```

Comment:
This is potentially very inefficient since we only ever have two elements. Also, the face->cell information already exists in the grid, available as
```
auto fc = faceCells(grid);
int c0 = fc(face_index, 0);
int c1 = fc(face_index, 1);
```
Here c0 or c1 can be -1, indicating a boundary.

You should change this to not recreate this information.